sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala

```diff
@@ -333,15 +333,14 @@ abstract class SparkPlan extends QueryPlan[SparkPlan] with Logging with Serializ
         // If we didn't find any rows after the first iteration, just try all partitions next.
         // Otherwise, interpolate the number of partitions we need to try, but overestimate it
         // by 50%.
-        if (buf.size == 0) {
-          numPartsToTry = totalParts - 1
+        if (buf.isEmpty) {
```

Comment:
Yeah, this is actually taken from `RDD.take()`. Your approach is closer to what that does, but I think we should actually make them operate as similarly as possible unless there's a reason not to. Could you compare and sync them up further?